### PR TITLE
[wip] Windowing and multiple-lookup functions

### PR DESCRIPTION
**Pull request by apendleton**
Carmen looks for matches anywhere in a given query in each index. At present it does this by finding all the subqueries on the JS side, and looking each of them up one by one in the phrase cache. This leads to a couple of kinds of duplicated effort:
* if the same token occurs in multiple phrases (as will typically be the case), we perform the same spelling-correction/variant-identification multiple times
* we consider multiple phrases of different lengths that start at the same position, leading to overlapping walks of the same parts of the phrase graph

The goal of this PR is to introduce ways to avoid this duplicated effort:
- [x] At a low level, introduce a variant of our existing phrase-graph walker that can start at a given position and try to walk to the longest phrase that starts at that position, emitting any valid states is passes along the way -- this means for windowed search, we only have to perform a phrase walk for each possible start position, rather than for each (start, end) pair
- [x] At a higher level, introduce a new method that takes a single query (= token list), and explores all substrings; for carmen, this should get us what we need for non-address indexes
- [x] Address indexes do funkier stuff where they rearrange input substrings to attempt to canonicalize the position of the house number as always at the start of the subquery; as a consequence, the set of subqueries carmen tries to look up in address indexes will not all be substrings of a single original query string. As a consequence, introduce an alternative to the windowed lookup above that just takes a list of arbitrary strings and looks all of them up even if they don't form windows in a single string, but attempting to realize work sharing where possible:
    - [x] identify all the unique tokens across all the submitted words and spelling-correct each one just once
    - [x] recognize that while overlap between phrases is not guaranteed, it will in practice often occur, and collapse together clusters of phrases that are prefixes of one another to reduce phrase walks, taking advantage of the same recursive phrase walker introduced above to walk overlapping phrases in concert

I expect this multi-search option to be slower than a windowed search when both are possible, but still faster than a bunch of individual lookups, so we should use this one when necessary and the windowed option when possible.

Most of what's left to do revolves around testing/benching. This code is only cursorily tested. It needs:
- [x] phrase-graph level tests; #39 lays foundations that should make this easier, but they haven't been done yet
- [x] more than ~1 glue-module-level unit test per function
- [x] additions to the exhaustive/fuzz-testing infrastructure so we can be confident about corner cases:
    - [x] these tests should simulate queries part of which match within the given index and part of which doesn't (and is instead, e.g., a city/state/zip), and probably also instances where more than one match in a given index occurs within the query (as happens when the terminal token is a postcode that also matches as a house number)
    - [x] we should test that windowed search and multi-search produce the same results when handed equivalent queries (i.e., a multi-search that just passes in all the windows)
    - [x] we should test that a single multi-search and multiple individual fuzzy searches produce the same results
- [x] benches that:
    - [x] establish the relative speed of single-phrase fuzzy vs. windowed vs. multi where all are equivalent
    - ~establish the relative speed of fuzzy vs. multi where both are equivalent, outside of the context where windowing is applicable (e.g., with address-style rearrangement of subqueries) -- this will likely require a Rust-side rough reimplementation (doesn't have to be perfect edge-case-wise, just representative) of the carmen address rearranging logic~ bench results so far suggest to me that this probably isn't necessary, and is a pain

Also would in general love some more eyes on the code.

### CODE REVIEW

**Comment by apendleton** [comment]
I should also say: I factored some code out of the existing fuzzy-match functions to reuse in the new functions, so that's what `get_terminal_word_possibilities` and `get_nonterminal_word_possibilities` are.

File: src/glue/mod.rs [hunk 1]
```diff
@@ -420,6 +438,240 @@ impl FuzzyPhraseSet {
...
+                let last_idx = phrase.len() - 1;
+                for word in phrase[..last_idx].iter() {
+                    let word = word.as_ref();
+                    if let hash_map::Entry::Vacant(entry) = all_words.entry((word, false)) {
+                        entry.insert(
+                            self.get_nonterminal_word_possibilities(word, edit_distance)?
+                                .unwrap_or_else(|| Vec::with_capacity(0))
+                        );
+                    }
+                }
```
Comment (boblannon) [inline src/glue/mod.rs]: any reason not to use `or_insert`? like

```
all_words.entry((word, false)).or_insert()
```

[eg in docs](https://doc.rust-lang.org/std/collections/struct.HashMap.html#examples-16)

Comment (boblannon) [reply src/glue/mod.rs]: err, i guess it would be `or_insert_with()` as in https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with

Comment (apendleton) [reply src/glue/mod.rs]: It's because of the `?` operator in `self.get_nonterminal_word_possibilities(word, edit_distance)?`. If you use the `?` inside the closure, it's an early return from the _closure_ rather than from the wrapping function, and the closure can't have a return type of Result, it has to have a return type of the thing you want to insert, so there doesn't seem to be a way to propagate the error state using `or_insert_with`. (And yeah, plain `or_insert` means you have to construct the thing you would insert even if you don't actually need it.)

Comment (boblannon) [reply src/glue/mod.rs]: ah gotcha.

File: src/glue/mod.rs [hunk 2]
```diff
@@ -420,6 +438,240 @@ impl FuzzyPhraseSet {
...
+                    item.1 ||
+                        peek.0.len() <= item.0.len() ||
+                        &peek.0[..item.0.len()] != item.0
+                },
+            };
+            if done_with_group {
+                collapsed.insert(item.2, group);
+                group = Vec::new();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (apendleton) [reply src/glue/mod.rs]: Yeah, this should be guaranteed by the sort, and there are corner cases where it won't be perfect (e.g., there's duplication of effort if we try and look up `[["A", "B", "C"], ["A", "B", "C", "D"], ["A", "B", "C", "E"]]` -- we'll duplicatively traverse ABC for the last one), but worst case scenario it fails to collapse some results and does extra work; it shouldn't be possible for it to produce results that are incorrect.

Good call on adding a note though.

**Comment by boblannon** [comment]
FYI: read through the changes and feel like i understand them. In the way of note-taking, I opened #41 with some additional documentation.

@apendleton If you'd like to move on to other things, I would feel comfortable writing tests/benches for what's been done so far.

**Comment by apendleton** [comment]
@boblannon mostly I'm eager to get this out the door, and the best way to get that to happen might well be to divvy it up rather than jump off, but when you pair in the morning, feel free to grab whatever, and if I end up working tomorrow I'll either grab some of what's left or PR the node-bindings side.

**Comment by apendleton** [comment]
**tl/dr:** both multi-lookup strategies help a ton on performance, with automatic windowing unsurprisingly being the better performer of the two, though by less of a margin than I'd have thought

I wrote some benchmarks that simulate a real-world multi-lookup context for addresses by building sample phrases consisting of an address from our test set with a single introduced typo, plus a city, state, and zip (city and state from newly added sample files, zip by randomly generating a 5-digit number), and then attempting to fuzzy-match all substrings using the following strategies:
* enumerate all the subqueries into a vector and call `fuzzy_match` on each
* enumerate all the subqueries and call `fuzzy_match_multi` on all of them as a group
* call `fuzzy_match_windows` on the token list

Benchmarks:
```
glue/fuzzy_match_complex_manual
                        time:   [704.00 us 714.60 us 725.87 us]
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low severe
  2 (2.00%) low mild
  3 (3.00%) high mild
glue/fuzzy_match_complex_multi
                        time:   [98.679 us 100.76 us 102.82 us]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe
glue/fuzzy_match_complex_windows
                        time:   [82.738 us 84.778 us 87.241 us]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

So, supposing 100 indexes, this makes for about 8-10ms of total time spent on lookups (which I think we can live with), down from ~70ms (not accounting for any node->rust overhead, of course)

Also: I'm having second thoughts about whether or not a version that does the number-rearranging thing is worth writing. It would be a fair bit of work, we have no suitable test data identified, and I'm not sure how revelatory it would be given my increasing confidence that most of the compute time is spent on spelling correction anyway. Curious about people's thoughts, though.

**Comment by boblannon** [comment]
> phrase-graph level tests; #39 lays foundations that should make this easier, but they haven't been done yet

#45 is meant to add these
